Hibernate ORM, Server configuration, Websockets (work in progress)

FakeEmperor opened the pull request:
Добрый день.

В нашем несмотря на количество коммитов, в нашем пулл-реквесте, изменений не так много:
- Появился контекст (Context). Теперь его можно получать в различных сервлетах для доступа к сервисам
- Вместо JDBC теперь используется Hibernate. Код работы с базой покрыт тестами.
- Сервер конфигурируется из файла. Под конфигурацией имеется в виду пока лишь выбор порта и хоста для сервера.
- Исправлены настройки log4j2
- Появились зачатки вебсокетов. Эхо работает и оповещает о событиях на вебсокете.
- Небольшой рефактор кода и перемещения некоторых классов в другие пакеты для наглядности.
- Статика вынесена в отдельный проект

Igor1711 committed:
igor-commit2

Igor1711 committed:
account_commit

Igor1711 committed:
accountcommit

Igor1711 committed:
accountcommit

Igor1711 committed:
accountcommit

Igor1711 committed:
accountcommit

Igor1711 committed:
password-hash

Igor1711 committed:
password-hash

FakeEmperor committed:
Fixed inspections

FakeEmperor committed:
Fixed inspections
Actually this commit contains all the inspections
: Name usages
: Spaces, brackets, commas
: CodeStyle

Igor1711 committed:
acccommit

Igor1711 committed:
new commit

Igor1711 committed:
new commit

Igor1711 committed:
usercommit

Igor1711 committed:
new commit

Igor1711 committed:
usercommit

FakeEmperor committed:
Merge pull request #7 from Igor1711/master

Merging #7 :+1:

Igor1711 committed:
checkerror

FakeEmperor committed:
Revert "account2"

FakeEmperor committed:
Merge pull request #8 from SlaykovskiyHDNANO/revert-7-master

Revert "account2"

Igor1711 committed:
commit

Igor1711 committed:
1errf

Igor1711 committed:
hibernate1

Igor1711 committed:
hibernate-commit

Igor1711 committed:
hibernate-commit2

Igor1711 committed:
bin-validation

Igor1711 committed:
optimise_validation_remove_logger_errors

Igor1711 committed:
optimise_validation_remove_logger_errors

Igor1711 committed:
add userservlet tests

Igor1711 committed:
add userservlet tests

FakeEmperor committed:
[ WIP ] Overhauling project code

FakeEmperor committed:
[ WIP ] Merge preparations

FakeEmperor committed:
[WIP] Minor fixes, added property to fetch configurations from /cfg folder

Igor1711 committed:
validation in accountservice

Igor1711 committed:
database_fixed

Igor1711 committed:
database_fixed

Igor1711 committed:
database_fixed1

Igor1711 committed:
database_configuration

Igor1711 committed:
database_configuration_in_xml

Igor1711 committed:
database_configuration_in_xml_without_frontend

Igor1711 committed:
database_configuration_in_xml_without_frontend

FakeEmperor committed:
Merge branch 'master' into temp_merge

FakeEmperor committed:
[ WIP ] Changed AccountService, and made cosmetic changes for code inspections

FakeEmperor committed:
Merge pull request #12 from SlaykovskiyHDNANO/igor

FakeEmperor committed:
[+] Hibernate database and tests

FakeEmperor committed:
fix: Not staged changes

FakeEmperor committed:
Fixes and minor update
[+] Added Context - see Main.java and RestAppV1 for usage.
fix: DBAccountService - used query without transaction
[REFACTOR] Removed some comments from RestAppV1 and Main

Igor1711 committed:
hibernate_validation

Igor1711 committed:
hibernate_validation

FakeEmperor committed:
[WIP] Websockets service, server configuration

KSolovyev commented:
А зачем второе ревью?

FakeEmperor commented:
Первый пулл реквест на ревью отдавал Игорь. Так получилось, что в какой-то момент мы начали параллельно делать одну и ту же фичу – датабазный AccountService. Игорь решил отправить отправить код на ревью, так как к тому времени другого кода ещё не было. Отличий достаточно много:
- Вместо JDBC используется Hibernate. У нас игра ещё в зачаточном состоянии, и вряд ли будет множество игроков (хэх). Так как проблем помимо написания SQL-кода у нас хватает, я думаю, что Hibernate в данном случае больше подходит, нежели JDBC.
- Добавился контекст – сервлеты (игровые и моделей) теперь используют контекст для доступа к сервисам, а сами сервисы запускаются в главном потоке.
- Конфигурация сервера стала менее тяжеловесной как в коде, так и в самом файле конфигурации.
- Конфиги Log4j2 стали наконец работать. До этого был конфиг формата старой версии.
- Появились зачатки игровых сервисов и сервиса обмена сообщениями.

Ветка игоря находится в его репозитории. Я попросил его закрыть пулл-реквест, так как тот код в той ветке уже неактуален. Думаю, в ближайшее время он это сделает.

Igor1711 committed:
sessionServletChecked

KSolovyev commented on src/main/java/db/services/impl/DBAccountServiceImpl.java:
```diff
@@ -0,0 +1,121 @@
+    @SuppressWarnings("unchecked")
+    @Override
+    public Collection<User> getUsers() {
+        final Session session = DBSessionFactory.getCurrentSession();
```
Любая работа с базой должна выглядеть примерно так:

```
        Transaction tx = null;
        try {
            final Session session = factory.openSession()
            tx = session.beginTransaction();
            final UsersDAO dao = new UsersDAO(session);
            return dao.getAllUsers();
        } catch (HibernateException ex) {
          if(tx != null && 
                  (tx.getStatus() == TransactionStatus.ACTIVE
                  || tx.getStatus() == TransactionStatus.MARKED_ROLLBACK)) {
              tx.rollback();
          }
            ex.printStackTrace();
            throw new DatabaseException("Unnable to get all users", ex);
        }
```

В месте про throw new DatabaseException("Unnable to get all users", ex); можно придумать что-то свое. Главное вы должны как-то сообщить пользователю, что не получилось, и, если необходимо, отличать случай ошибки базы данных от пустого результата

KSolovyev commented on src/main/java/db/services/impl/DBSessionFactory.java:
```diff
@@ -0,0 +1,42 @@
+ * package: db.services.impl
+ */
+public class DBSessionFactory {
+    static SessionFactory sessionFactory;
```
переделайте из синглтона в сервис. И просто прокидывайте его как зависимость

FakeEmperor committed:
[WIP] Pull request fixes and MessagingService

FakeEmperor committed:
New application configs, Jersey ResourceConfig instead Application and new dependencies

FakeEmperor committed:
Fixes for pull request and new methods
[+] Added transaction rollback and Executor to simplify this
[+] Added new method: getScores() and tests for it
[+] New model - UserScore for method above
[+] Fixed tests to comply with new exceptions
[WIP] Now uses rmi.AccessException to throw them. TODO: use custom
exceptions
[+] Implemented changeUser() method. Added test for it
[+] New AccountDAO class

FakeEmperor committed:
[WIP] Messaging system, fixes in servlets
[~] Due to the new method of booting our RESTful resources, it was
necessary to add @Inject annotation
[fix] ServletContext was null in MessagingServlet

Igor1711 committed:
fd

Igor1711 committed:
Merge branch 'master' of https://github.com/SlaykovskiyHDNANO/2016-02-Lime-Pineapple into igor

Igor1711 committed:
error

Igor1711 committed:
GAMEROOM

Igor1711 committed:
cardeffects

Igor1711 committed:
gameserver

KSolovyev commented:
Как дела? У вас большие проблемы с количеством баллов. Попробуйте закрыть хотя бы первые 2 РК

FakeEmperor commented:
Добрый день.
Да, хотим сдать как можно больше кода на ревью к как можно большему количеству РК :) А по теории: можно ли вам написать, чтобы получить вопросы к РК?
По текущему пулл-реквесту: предпоследний коммит, вроде бы, правит те недостатки, что были отмечены в коде.

FakeEmperor committed:
Merge pull request #14 from SlaykovskiyHDNANO/feature_hibernate_orm

Hibernate ORM, Server configuration

FakeEmperor committed:
Merge pull request #15 from SlaykovskiyHDNANO/feature_websockets

[WIP] Feature websockets

FakeEmperor committed:
Uncommited changes: CORS filter, a
little refactoring
[+] Added CORS filter: it is used to allow RESTful services on different
from back-end server domains (like our
front-end server) to send REST requests to our server
[+] Added final modifier to variable inside SessionServlet.addIdFromRequest
[~] Changed PUT to POST in addSession method
[+] Refactored dependency placing inside pom.xml

FakeEmperor committed:
Merge branch 'igor'
Resolved conflicts with priority to Igor's branch

Igor1711 committed:
commit

Igor1711 committed:
Merge branch 'master' of https://github.com/SlaykovskiyHDNANO/2016-02-Lime-Pineapple into igor

Igor1711 committed:
GameRoomServices

KSolovyev commented on src/main/java/db/services/AccountService.java:
```diff
@@ -2,16 +2,33 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
+import java.rmi.AccessException;
```
rmi - это Remote method invocation. Если хотите использовать исключения, то пишите здесь свое и наследуйтесь от Exception

KSolovyev commented:
Если вы деплоитесь на свой хост, то стоит удалить пароль к базе из гита

KSolovyev commented:
тест testSmokeHibernateConnection и testSmokeHibernateSession используют боевой конфиг
testServletGetAll - сломан

FakeEmperor committed:
Tests: updated tests
[+] Added tests ((de)serialization) for Message
[+] Fixed testServletGetAll test
[+] Fixed: because we're using now our own DatabaseException, fixes are necessary for try {...} catch {...} blocks

FakeEmperor committed:
Configs: updated configs and dependencies
[+] Newly added DB models are in configs
[+] Fixed DB names

FakeEmperor committed:
DB: Models, Exceptions
[ + ] New CardModel - a representation of a Card in DB
[ + ] Changed CardType to conform new cards
[ + ] New EffectModel - a representation of an Effect (a modifier to a card) in DB

FakeEmperor committed:
Final Message system implementation
Fully implemented message system is rolled
 [ + ] MessageDeserializer - used to deserialize JSON message to Message class (or more concrete class)
 [ + ] We use Fibers for Multithreading and lightweight message dispatching
 [ + ] For message responses, we use async writes with Futures

FakeEmperor committed:
fix: Using new DatabaseException inside servlets to handle DB errors

FakeEmperor committed:
:rooster: :lollipop: Beta for game mechanics and changes in Main
Main:
[ + ] Now creating GameEngineService
game.services:
[ + ] GameEngineService: almost done. Need to implement core logic
[ + ] GameCardService is responsible for handling CardModels and fetching them from database. It also generates hands for Players
[ + ] GameFieldFactoryService is responsible for creation Field
[ + ] GameField - is a container, where things are happening
[ + ] GameFieldRow - is a row in a GameField
[ + ] GameRoom - all needed methods are done. Yippee!
[ ? ] PlayingUser - maybe he's useless?
[ + ] Card - is a base class for all cards that will follow. Stores card effects and current card state
[ + ] MatchmakingService - a stub for later implementation

FakeEmperor committed:
fix: Unstaged commits

FakeEmperor committed:
Merge branch 'feature_game' into feature_websockets

FakeEmperor commented:
Добрый день. 
Пофиксил тесты, должны работать. Правда поменял ещё и конфиг базы данных.
Хорошо, удалим, когда будем деплоить на хост.
Появилась более конкретная механика, полностью реализованная система сообщений.

KSolovyev commented:
Играется ли у вас игра? Есть ли версия, чтобы можно было ее раздеплоить? Если можно поиграть, то я вам поставлю баллы за работу на 3-ем РК. А вечером еще посмотрю как у вас дела с ревью

FakeEmperor committed:
[WIP] Game mechanics: CardService and more cards!

FakeEmperor commented:
Доброго вечера. Пока нет. Мы только начали склеивать вебсокеты вместе.
